LEARNER-36: Add new relic custom metrics to course home.

robrap opened the pull request:
## [LEARNER-36](https://openedx.atlassian.net/browse/LEARNER-36)

### Description

Adds New Relic custom metrics to Course Home page.

### Sandbox
- [X] [https://robrap.sandbox.edx.org](https://robrap.sandbox.edx.org)

### Testing
- [X] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @andy-armstrong 

### Post-review
- [x] Rebase and squash commits

robrap commented on openedx/core/lib/new_relic_utils.py:
```diff
@@ -0,0 +1,24 @@
+
+"""
+
+try:
```
@andy-armstrong: Due to this code, it seemed reasonable to make this a utility.  However, this means this would need to live in a core repo if we were to ever move Course Home to its own repo.

FYI: @jbarciauskas

andy-armstrong commented on openedx/features/course_experience/views/course_home.py:
```diff
@@ -35,6 +36,7 @@ def get(self, request, course_id):
         """
         course_key = CourseKey.from_string(course_id)
         course = get_course_with_access(request.user, 'load', course_key, check_if_enrolled=True)
+        add_new_relic_course_key_metrics(course_key)
```
I wonder if this could be centralized in ``CourseTabView``. It seems that it would be worth adding these metrics to all course tabs (note that not all course tabs currently subclass from this view, but it is the direction I'd like us to move as we convert tabs to fragments).

robrap replied:
Good call.  I'll look into that.

andy-armstrong commented on openedx/core/lib/new_relic_utils.py:
```diff
@@ -0,0 +1,24 @@
+"""
+Contains utilities for working with New Relic.
```
Nit: I don't love the idea of having a file dedicated to a particular product. Maybe rename it to analytics_utils.py?

robrap replied:
Let's discuss.  I don't know the right answer.  However, I know that when I was doing our platform inventory project I found it annoying to keep finding generically named files including a single non-generic method.  

The question is when we don't have a coherent library of utilities, and we just have a single utility method, what is the best way to have it found, understood, and potentially grow into a library.  I was leaning toward more specific to start and grouping into more generic files (or directories) as more gets added.  

But, unfortunately, copy/paste is the most likely way for this utility to get re-used, and not due to searching for it or seeing appropriate names.  

Anyway, it's an open question I have about how to get the most reuse from any utilities.  Ultimately we'll need searchable documentation, especially as some of these get moved to a separate repo.

andy-armstrong replied:
One idea I had last night is to make it a static method on ``CourseTabView``. That way it can be called easily from course tab views, but it allows views that haven't been converted to call it too. It also makes it clearer what it is: ``CourseTabView.add_new_relic_course_key_metrics``. What do you think?

robrap replied:
Sounds like a plan.

jbarciauskas commented:
I'd prefer "perf", "monitoring" or "instrumentation" (while we're
nitpicking :))

On Mon, Mar 27, 2017 at 5:10 PM, Andy Armstrong <notifications@github.com>
wrote:

> *@andy-armstrong* commented on this pull request.
>
> Looks good, @robrap <https://github.com/robrap>. I'll hold my thumbs
> until we've tested it on a sandbox.
> ------------------------------
>
> In openedx/features/course_experience/views/course_home.py
> <https://github.com/edx/edx-platform/pull/14774#discussion_r108280303>:
>
> > @@ -35,6 +36,7 @@ def get(self, request, course_id):
>          """
>          course_key = CourseKey.from_string(course_id)
>          course = get_course_with_access(request.user, 'load', course_key, check_if_enrolled=True)
> +        add_new_relic_course_key_metrics(course_key)
>
> I wonder if this could be centralized in CourseTabView. It seems that it
> would be worth adding these metrics to all course tabs (note that not all
> course tabs currently subclass from this view, but it is the direction I'd
> like us to move as we convert tabs to fragments).
> ------------------------------
>
> In openedx/core/lib/new_relic_utils.py
> <https://github.com/edx/edx-platform/pull/14774#discussion_r108280440>:
>
> > @@ -0,0 +1,24 @@
> +"""
> +Contains utilities for working with New Relic.
>
> Nit: I don't love the idea of having a file dedicated to a particular
> product. Maybe rename it to analytics_utils.py?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-platform/pull/14774#pullrequestreview-29302215>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAS-NZ4Z4YYdhstKCW7mn9l27YVnfZXGks5rqCXYgaJpZM4Mq0jQ>
> .
>



-- 
*Joel Barciauskas*
edX | Engineering Manager, Open edX
141 Portland Street, 8th floor
Cambridge, MA 02139
joel@edx.org

robrap commented on openedx/features/course_experience/views/course_home.py:
```diff
@@ -35,6 +36,7 @@ def get(self, request, course_id):
         """
         course_key = CourseKey.from_string(course_id)
         course = get_course_with_access(request.user, 'load', course_key, check_if_enrolled=True)
+        CourseTabView.add_new_relic_course_key_metrics(course_key)
```
@andy-armstrong I know this should be removed once this becomes a CourseTabView, but I guess we'll see who merges first. ;)

andy-armstrong replied:
Ah, I'd forgotten that that was only in my PR. Let's get yours merged first!

robrap commented:
@andy-armstrong I made the changes you requested, and I am building a sandbox now.

robrap commented:
Test NRQL: SELECT browserTransactionName, course_id, org FROM PageView SINCE 30 MINUTES AGO

[New Relic Insights link](https://insights.newrelic.com/accounts/1391042/explorer/events?eventType=PageView&facet=course_id&duration=1800000).

Results screenshot:
![newrelic-events-with-custom-metrics](https://cloud.githubusercontent.com/assets/14576445/24422870/02edd150-13c9-11e7-8fdb-108dd7a6459a.png)
FYI: @andy-armstrong

robrap commented:
@andy-armstrong: From the results, you can see we got these metrics on Discussions as well for free (assuming they weren't there before). :)

However, there are a couple of other places where it _may_ make sense to refactor to use this new call:
- [tasks.py](https://github.com/edx/edx-platform/blob/6a55ef7920e12d78e98d9a03603cc55558b08165/lms/djangoapps/grades/tasks.py#L106): Utility function from common (see earlier commit) would make more sense than where we moved it.  Using function would have the benefit of adding "org" metric.
- [module_render.py](https://github.com/edx/edx-platform/blob/6a55ef7920e12d78e98d9a03603cc55558b08165/lms/djangoapps/courseware/module_render.py#L975): Same issue with where the utility function lives.
- [middleware.py](https://github.com/edx/edx-platform/blob/630985b3cc69c46991472de3fb61d98bbb4bfbdc/openedx/core/djangoapps/contentserver/middleware.py#L93-L93): May not be worth refactoring because other metrics are also added.

FYI: @ormsbee

andy-armstrong commented on lms/djangoapps/courseware/views/views.py:
```diff
@@ -523,6 +528,20 @@ def render_to_standalone_html(self, request, fragment, course=None, tab=None, **
+    @staticmethod
+    def add_new_relic_course_key_metrics(course_key):
+        """
+        Initialize metrics (course_id and org) for New Relic from a course_key,
```
Nit: docstrings should start with a single line summary:

http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python_guidelines.html#docstrings

andy-armstrong commented on openedx/core/lib/monitoring_utils.py:
```diff
@@ -0,0 +1,39 @@
+"""
+Contains utilities for working with New Relic.
```
Nit: either call this file new_relic_utils or change the comment to be broader.

robrap commented:
@andy-armstrong: FYI: Rebase is complete.  I removed the changes from course_home.py since they were also in CourseTabView.  Also updated the comment.

Once tests are passing I will rebuild sandbox and verify the additional refactors, squash, and merge.

ormsbee commented:
FYI: @pwnage101

pwnage101 commented:
I'll just clarify that there is already one place where custom metrics reporting has been centralized: 
https://github.com/edx/edx-platform/blob/master/common/djangoapps/newrelic_custom_metrics/__init__.py

Ideally we wouldn't invite ambiguity by providing both of these modules:

* `openedx/core/lib/monitoring_utils.py`
* `common/djangoapps/newrelic_custom_metrics`

which seem to offer similar features.  As far as exclusive features,  newrelic_custom_metrics supports accumulation of numbers, whereas monitoring_utils.py supports strings.  It feels to me like we can just merge them?

pwnage101 commented:
Had a chat with @robrap and decided that the quickest way forward is to copy the contents of `monitoring_utils.py` into `common/djangoapps/newrelic_custom_metrics/__init__.py` so they at least live in the same place.  We also want to rename "newrelic_custom_metrics" to something else, and remove mentions of "newrelic" from docstrings wherever appropriate.

Just to be clear, functions in `__init__.py` are for public use, and may or may not utilize the middleware.  Direct calls to the newrelic agent from `__init__.py` are okay moving forward, IMO.

robrap commented:
It turns out that `openedx/core/djangoapps/monitoring`, but isn't doing anything related to monitoring.  

@ormsbee: How do you feel about this plan?
- Rename `openedx/core/djangoapps/monitoring` to `openedx/core/djangoapps/exceptions`.
- Move `common/djangoapps/newrelic_custom_metrics` to `openedx/core/djangoapps/monitoring_utls`.
- Move new `openedx/core/lib/monitoring_utils.py` code into `__init__.py` of `openedx/core/djangoapps/monitoring_utls`.

andy-armstrong commented:
@robrap that sounds good to me. Could we just kill the monitoring Django app and move the one method into a different Django app? I don't like having entire apps for just one method. Having said that, I can't find a natural home for it in any of the existing apps.

ormsbee commented:
@robrap: LGTM

robrap commented:
@andy-armstrong I'm not hearing a proposal, and I want to close this out.  I'm planning on going for this refactor soon.

andy-armstrong commented:
@robrap I didn't intend you to be blocked on my comment, it was more to provoke discussion. My recommendation would be to leave that Django app alone in this PR since you aren't changing it. If you do want to rename it, I don't like calling it exceptions because what else would go in it? It seems like it could fit into a broader scope of over-arching functionality to support the platform, so maybe ``base_platform`` or ``platform_utils`` or something like that. Then when there are future methods like this they could go into this app rather than creating yet another one-function-app.

andy-armstrong commented:
Some existing Django apps that could reasonably go into a base_platform app:

 - common/djangoapps/request_cache
 - openedx/core/djangoapps/header_control
 - openedx/core/djangoapps/session_inactivity_timeout
 - openedx/core/djangoapps/util (not really sure what this is)

robrap commented:
@andy-armstrong I'm struggling to close out this 1 point story and it keeps growing.  

To close this, we all agree I need to rename the app that is currently named `monitoring`, but no longer contains `monitoring`.  I personally don't like the idea of generically named one function apps. 

I DO agree with your thoughts around consolidation.  However, I think that should be a separate PR, that includes a README, and has a bit more agreement on naming and what to include, and actually includes some consolidation.  If people come to consensus tonight before I implement tomorrow, I'd be happy to implement. :)

FYI: @nasthagiri: Re: Taming the Monolith

andy-armstrong commented:
@robrap I think you should leave renaming the app to another PR and just finish this one point story. IMO it doesn't help to rename it to another name that doesn't make long term sense.

robrap commented:
jenkins run js

robrap commented:
jenkins run bokchoy

robrap commented:
@pwnage101 This is ready for review again.  Thanks.

pwnage101 commented on openedx/core/lib/monitoring_utils.py:
```diff
@@ -0,0 +1,40 @@
+All New Relic related utilities will properly skip execution in an environment
+that does not use New Relic.
+
+"""
```
Is this file used anymore?  I think you forgot to `git rm`?

robrap replied:
Good call.  My bad.  I just forgot to delete it.

pwnage101 commented on openedx/core/djangoapps/monitoring_utils/__init__.py:
```diff
@@ -0,0 +1,97 @@
+    accumulate(name, 1)
+
+
+def add_custom_metrics_for_course_key(course_key):
```
Do you think we could rename this from `add_` to `set_`?  Now that these are alongside the `aggregate` and `increment` functions which have "memory", I would like to emphasize that these functions do *not* have memory by calling them `set`ters.

pwnage101 commented on openedx/core/djangoapps/monitoring_utils/__init__.py:
```diff
@@ -0,0 +1,97 @@
+    newrelic.agent.add_custom_parameter('org', unicode(course_key.org))
+
+
+def add_custom_metric(key, value):
```
ditto

pwnage101 commented on openedx/core/djangoapps/monitoring_utils/__init__.py:
```diff
@@ -0,0 +1,97 @@
+
+def set_monitoring_transaction_name(name, group=None, priority=None):
+    """
+    Sets the transaction name for monitoring.
```
this one already has set in the name.
